translator/trace/jaeger/jaegerproto_to_traces.go

```diff
@@ -138,27 +140,39 @@ func translateJaegerVersionAttr(attrs pdata.AttributeMap) {
 	}
 }
 
-func jSpansToInternal(spans []*model.Span, dest pdata.SpanSlice) {
-	if len(spans) == 0 {
-		return
-	}
+func jSpansToInternal(spans []*model.Span) map[instrumentationLibrary]*pdata.InstrumentationLibrarySpans {
+	spansByLibrary := make(map[instrumentationLibrary]*pdata.InstrumentationLibrarySpans)
```

Comment:
As @bogdandrutu mentioned, maps are pretty inefficient. This is a hot spot for conversion and most of the times we will allocate a map with just one key. So it would be great to avoid allocating a map if possible. I think it would be acceptable if we break down the spans like this:
```
{jSpansWithLibrary1..., jSpansWithLibrary2..., jSpansWithLibrary1...} -> {otlpLibrary1Spans, otlpLibrary2Spans, otlpLibrary1Spans}
```
It should simplify things and do not require a map.

Reply:
@bogdandrutu any thought on this ^ ?

Reply:
That will duplicate library info on each span, increasing memory usage. Do we have performance tests or benchmarks for that?

Reply:
It may increase memory usage for the worst cases when we have many spans with different libraries jumping like spansLib1, spansLib2, spansLib1 etc. I don't know how realistic it that use case? I assume that in 99.9% cases it'll be just one library, right? Otherwise we might also consider sorting by library before, but not sure if we need to optimize for that.
I don't think we have existing performance tests for this, but it might worth adding a benchmark test to this particular function along with the change.

Reply:
Mm, probably I misunderstood your proposal. If I remove the map, then I will need to convert every `model.Span` to a pair of `(instrumentationLibrary;pdata.Span)`, right? Which means having a separate copy of `instrumentationLibrary` for every span.

Reply:
My other concern is the scope of this PR, which is again exploding. I just wanted to implement recent changes in OTLP->Jaeger translation regarding `InstrumentationLibrary`. It already grew into including Jaeger->OTLP translation which is much more complicated. If I have also to add a benchmark test for this, that will blow my time budget for this task too much.

Reply:
I understand, but and the same time we shouldn't be introducing performance issues with this change just to merge it sooner. You don't need to return pairs, probably nothing should be returned as it was before the change, you can just append to an `InstrumentationLibrary` slice provided to this function instead.

Reply:
I afraid I don't understand you. This function takes an array of `model.Span`. Before, it converted them one by one and added them all into a slice of `pdata.Span`. Now we have to extract `pdata.InstrumentationLibrary` from every `model.Span` _in addition_ to `pdata.Span`. Then all those spans have to be grouped by `pdata.InstrumentationLibrary` into a collection of `pdata.InstrumentationLibrarySpans`. Each instance of `pdata.InstrumentationLibrarySpans` contains `pdata.InstrumentationLibrary` and a slice of `pdata.Span` created by that `pdata.InstrumentationLibrary`. Without the map of `pdata.InstrumentationLibrary -> collection of pdata.Span` I don't know how perform "group by" operation.

Reply:
I suggest the following:
1. Sort jaeger spans slice by instrumentation library (optional);
2. Iterate over the jaeger spans slice and build `pdata.InstrumentationLibrarySpans` structure on the fly: 
      - if jaeger instrumentation library is the same as previous: add the span to the current `pdata.InstrumentationLibrarySpans->Spans` slice.
      - otherwise: add new item to `pdata.InstrumentationLibrarySpans` slice.

My initial point that I wouldn't care about duplicated instrumentation library data, that's why I think `1` is optional.

Does it make sense?

Reply:
A, yes, it does :) If you think this extra sorting is faster overall than using a map, I will try that tomorrow.

Reply:
Yes, I think it must be faster, especially assuming that in vast majority cases it'll be just one library.

I'd say it's still an open question if we need the sorting at all. There are plans to remove this instrumentation library wrapping sometime in future. I'd like to hear what others think about it. cc @bogdandrutu

Reply:
Indeed, maps are expensive but if we do what you suggest @dmitryax we will allocate tons of data, so in this particular case I tend to agree with the map approach.

We will allocate a new "instrumentation library object", a new slice of spans, and a new "entry to the resource slice of instr lib". So 3 extra allocations which I think is more than the overhead of using maps.

Reply:
Yeah, that's the case with no sorting and different instrumentation libraries. I still think this is going to be unnecessary map allocation just for one key most of the times, but probably we should not predict it too much and assume the worth case. Happy to close the discussion if there are no other comments